<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>2023-09-07: UQ PUG 1 – UQ Python User Group</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../../styles.css">
</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item">Session Notebooks</li><li class="breadcrumb-item"><a href="../../Session_notebooks/Past - 2023/1-Sep23.html">Past - 2023</a></li><li class="breadcrumb-item"><a href="../../Session_notebooks/Past - 2023/1-Sep23.html">2023-09-07: UQ PUG 1</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../../">UQ Python User Group</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">General information</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Session Notebooks</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="false">
 <span class="menu-text">Current - 2024</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth2 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../Session_notebooks/Current - 2024/12-Oct24.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">2024 October 3<sup>rd</sup> – UQ PUG 12</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../Session_notebooks/Current - 2024/13-Nov24.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">2024 November 7<sup>th</sup> – UQ PUG 13</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../Session_notebooks/Current - 2024/5-Feb24.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">2024-2-7: UQ PUG 5</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../Session_notebooks/Current - 2024/6-Mar24.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">2024-3-7: UQ PUG 6</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../Session_notebooks/Current - 2024/7-Apr24.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">2024-5-11: UQ PUG 7</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../Session_notebooks/Current - 2024/8-May24.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">2024 May 2<sup>nd</sup> – UQ PUG 8</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../Session_notebooks/Current - 2024/9-Jul24.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">2024 July 4<sup>th</sup> – UQ PUG 9</span></a>
  </div>
</li>
      </ul>
  </li>
          <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true">
 <span class="menu-text">Past - 2023</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth2 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../Session_notebooks/Past - 2023/1-Sep23.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">2023-09-07: UQ PUG 1</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../Session_notebooks/Past - 2023/2-Oct23.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">2023-10-05: UQ PUG 2</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../Session_notebooks/Past - 2023/3-Nov23.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">2023-11-02: UQ PUG 3</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../Session_notebooks/Past - 2023/4-Dec23.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">2023-12-07: UQ PUG 4</span></a>
  </div>
</li>
      </ul>
  </li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#this-months-presentationwelcome-to-our-first-python-user-group-gathering-this-month-luke-and-cameron-give-an-overview-of-the-group-our-vision-and-noteable-the-interactive-collaborative-notebook-platform-which-can-run-markdown-python-r-and-sql." id="toc-this-months-presentationwelcome-to-our-first-python-user-group-gathering-this-month-luke-and-cameron-give-an-overview-of-the-group-our-vision-and-noteable-the-interactive-collaborative-notebook-platform-which-can-run-markdown-python-r-and-sql." class="nav-link active" data-scroll-target="#this-months-presentationwelcome-to-our-first-python-user-group-gathering-this-month-luke-and-cameron-give-an-overview-of-the-group-our-vision-and-noteable-the-interactive-collaborative-notebook-platform-which-can-run-markdown-python-r-and-sql.">This month’s presentationWelcome to our first Python User Group gathering! This month, Luke and Cameron give an overview of the group, our vision and Noteable, the interactive collaborative notebook platform which can run markdown, Python, R and SQL.</a></li>
  <li><a href="#introduce-yourself-whats-your-name-where-are-you-from-why-are-you-here-luke-library-learn-nick-library-learn-valentina-urrutia-guada-library-learn-nida-school-of-molecular-biology-1-understanding-the-logic-in-some-codes-for-bioinformatics-and-metagenomics-application-2-learn-how-to-visualize-data-in-python-that-is-hard-to-do-in-excel-cameron-library-learn-sam-hames-school-of-languages-and-cultures-communitynikhil-school-of-eecs-learn-paul-vrbik-eecs-support-jason-dail-senv-learn-annie-nguyen-senv-learn" id="toc-introduce-yourself-whats-your-name-where-are-you-from-why-are-you-here-luke-library-learn-nick-library-learn-valentina-urrutia-guada-library-learn-nida-school-of-molecular-biology-1-understanding-the-logic-in-some-codes-for-bioinformatics-and-metagenomics-application-2-learn-how-to-visualize-data-in-python-that-is-hard-to-do-in-excel-cameron-library-learn-sam-hames-school-of-languages-and-cultures-communitynikhil-school-of-eecs-learn-paul-vrbik-eecs-support-jason-dail-senv-learn-annie-nguyen-senv-learn" class="nav-link" data-scroll-target="#introduce-yourself-whats-your-name-where-are-you-from-why-are-you-here-luke-library-learn-nick-library-learn-valentina-urrutia-guada-library-learn-nida-school-of-molecular-biology-1-understanding-the-logic-in-some-codes-for-bioinformatics-and-metagenomics-application-2-learn-how-to-visualize-data-in-python-that-is-hard-to-do-in-excel-cameron-library-learn-sam-hames-school-of-languages-and-cultures-communitynikhil-school-of-eecs-learn-paul-vrbik-eecs-support-jason-dail-senv-learn-annie-nguyen-senv-learn">Introduce yourself| What’s your name? | Where are you from? | Why are you here? | | — | — | — ||Luke |Library |Learn ||Nick |Library |Learn ||Valentina Urrutia Guada |Library |Learn ||Nida |School of Molecular Biology |(1) Understanding the logic (in some codes) for bioinformatics and metagenomics application, (2) Learn how to visualize data in python that is hard to do in excel || Cameron | Library | Learn ||Sam Hames | School of Languages and Cultures| Community||Nikhil |School of EECS | Learn|| Paul Vrbik | EECS | Support || Jason Dail| SENV | Learn ||Annie Nguyen |SENV | Learn :)|</a></li>
  <li><a href="#research-toolshere-are-a-few-links-we-shared-around-mostly-from-jason.1.-to-assist-with-academic-citations1.-httpsresearchrabbitapp.comhome-helpful-when-making-research-collections-mapping-concepts-and-looking-at-the-linkages-between-references1.-httpsconsensus.app1.-httpschat.openai.com1.-httpsarticle-summarizer.scholarcy.comsummarizer1.-httpstypeset.io-ai-assistant-for-reading-and-understanding-papers1.-httpswww.listendata.com202303how-to-run-chatgpt-inside-excel.html-excel-extension-for-chat-gpt-how-to1.-eecs-tutor-help-httpseecs.uq.edu.aucurrent-studentseecs-learning-centre-tutors" id="toc-research-toolshere-are-a-few-links-we-shared-around-mostly-from-jason.1.-to-assist-with-academic-citations1.-httpsresearchrabbitapp.comhome-helpful-when-making-research-collections-mapping-concepts-and-looking-at-the-linkages-between-references1.-httpsconsensus.app1.-httpschat.openai.com1.-httpsarticle-summarizer.scholarcy.comsummarizer1.-httpstypeset.io-ai-assistant-for-reading-and-understanding-papers1.-httpswww.listendata.com202303how-to-run-chatgpt-inside-excel.html-excel-extension-for-chat-gpt-how-to1.-eecs-tutor-help-httpseecs.uq.edu.aucurrent-studentseecs-learning-centre-tutors" class="nav-link" data-scroll-target="#research-toolshere-are-a-few-links-we-shared-around-mostly-from-jason.1.-to-assist-with-academic-citations1.-httpsresearchrabbitapp.comhome-helpful-when-making-research-collections-mapping-concepts-and-looking-at-the-linkages-between-references1.-httpsconsensus.app1.-httpschat.openai.com1.-httpsarticle-summarizer.scholarcy.comsummarizer1.-httpstypeset.io-ai-assistant-for-reading-and-understanding-papers1.-httpswww.listendata.com202303how-to-run-chatgpt-inside-excel.html-excel-extension-for-chat-gpt-how-to1.-eecs-tutor-help-httpseecs.uq.edu.aucurrent-studentseecs-learning-centre-tutors">Research toolsHere are a few links we shared around, mostly from Jason.1. to assist with academic citations1. https://researchrabbitapp.com/home helpful when making research collections, mapping concepts, and looking at the linkages between references1. https://consensus.app/1. https://chat.openai.com/1. https://article-summarizer.scholarcy.com/summarizer1. https://typeset.io/ AI assistant for reading and understanding papers1. https://www.listendata.com/2023/03/how-to-run-chatgpt-inside-excel.html (excel extension for chat GPT how-to)1. EECS tutor help: https://eecs.uq.edu.au/current-students/eecs-learning-centre-tutors</a></li>
  <li><a href="#questionsif-you-have-any-python-questions-youd-like-to-explore-with-the-group-please-put-them-in-a-markdown-cell-with-any-code-youd-like-us-to-run-in-a-python-cell." id="toc-questionsif-you-have-any-python-questions-youd-like-to-explore-with-the-group-please-put-them-in-a-markdown-cell-with-any-code-youd-like-us-to-run-in-a-python-cell." class="nav-link" data-scroll-target="#questionsif-you-have-any-python-questions-youd-like-to-explore-with-the-group-please-put-them-in-a-markdown-cell-with-any-code-youd-like-us-to-run-in-a-python-cell.">QuestionsIf you have any Python questions you’d like to explore with the group, please put them in a markdown cell, with any code you’d like us to run in a Python cell.</a>
  <ul class="collapse">
  <li><a href="#question-1---finding-substrings-for-covid-sequencing---note-that-the-formatting-has-not-transferred-correctly-here-nidanida-has-a-problem-where-she-needs-to-identify-specific-substrings-from-a-large-sequence-of-characters-dna-sequence.-her-code-is-belowi-just-found-out-that-the-code-covid_seq-getsequencemn908947-genbank-dna_alphabet-is-actually-trying-to-get-a-dna-sequence-of-covid-from-genbank-will-give-output-of-a-string-containing-29900-characters-that-we-can-actually-see-by-clicking-on-the-ebi-website-below-it-is-use-a-function-called-getsequence-with-codedef-getsequenceentryid-dbname-uniprotkb-alphabet-protein_alphabet-format-fasta-debug-bool-true-if-not-isinstanceentryid-str-entryid-entryid.decodeutf-8-url-httpwww.ebi.ac.uktoolsdbfetchdbfetchstylerawdb-dbname-format-format-id-entryid-try-if-debug-printdebug-querying-url-0.formaturl-data-urllib.request.urlopenurl.read-if-format-fasta-return-readfastastringdata.decodeutf-8-alphabet0-else-return-data.decodeutf-8-except-urllib.error.httperror-as-ex-raise-runtimeerrorex.readthis-function-retrieves-a-single-entry-from-a-database-entryid-id-for-entry-e.g.mn908947-dbname-name-of-database-e.g.-genbankonce-we-got-that-dna-string-consisting-of-only-4-types-of-characters-actg-a-biologist-will-translate-them-into-aminoacid-or-protein-string-consisting-20-types-of-characters-asterisk-please-see-httpswww.hgvs.orgmutnomencodon.html.-well-translate-them-using-a-dictionary-code-not-shown-and-then-well-split-them-based-on-so-that-we-can-generate-an-output-of-a-list-of-strings-that-is-stored-in-a-variable-called-protseq-see-httpsdocs.google.comdocumentd1r22igmfe9i1tyalpk5zsoikv-on6xzvuq-c87xdmilgedituspsharing-finally-from-that-i-need-to-find-how-many-strings-inside-of-that-protseq-that-meet-these-criteria1-start-with-m2-end-with-3-has-length-of-100-charactersso-my-question-is-actuallyhow-to-understand-the-logic-behind-the-below-code-that-is-said-to-be-able-to-do-that-jobcheck-first-occurence-of-m-in-each-string-of-that-protseq-listwhere_m_in_protseq_1st-string-protseq0.findmprintwhere_m_in_protseq_1st-stringprintlenprotseq0code-to-check-whether-each-string-in-that-protseq-start-with-m-and-100cnt-0for-i-in-protseqif-leni---i.findm-100cnt-1printcnt-for-seq-in-protseqm_pos-seq.findm-m_end_seq-seqm_posif-lenm_end_seq-100printm_end_seqprintlenm_end_seq-the-answer-for-this-problem-should-be-8-strings-inside-18-members-of-that-protseq-list-will-meet-that-criteria-so-if-we-can-end-up-getting-8-from-that-code-we-are-correct-but-i-just-hope-that-i-can-get-it-correct-and-understand-how-the-code-worksimportance-from-that-8-strings-we-can-try-1-of-them-in-a-real-protein-database-called-uniprot-kb-and-know-what-part-of-covid-body-that-is-likely-to-interact-with-human-and-causing-diseasethank-you-very-much" id="toc-question-1---finding-substrings-for-covid-sequencing---note-that-the-formatting-has-not-transferred-correctly-here-nidanida-has-a-problem-where-she-needs-to-identify-specific-substrings-from-a-large-sequence-of-characters-dna-sequence.-her-code-is-belowi-just-found-out-that-the-code-covid_seq-getsequencemn908947-genbank-dna_alphabet-is-actually-trying-to-get-a-dna-sequence-of-covid-from-genbank-will-give-output-of-a-string-containing-29900-characters-that-we-can-actually-see-by-clicking-on-the-ebi-website-below-it-is-use-a-function-called-getsequence-with-codedef-getsequenceentryid-dbname-uniprotkb-alphabet-protein_alphabet-format-fasta-debug-bool-true-if-not-isinstanceentryid-str-entryid-entryid.decodeutf-8-url-httpwww.ebi.ac.uktoolsdbfetchdbfetchstylerawdb-dbname-format-format-id-entryid-try-if-debug-printdebug-querying-url-0.formaturl-data-urllib.request.urlopenurl.read-if-format-fasta-return-readfastastringdata.decodeutf-8-alphabet0-else-return-data.decodeutf-8-except-urllib.error.httperror-as-ex-raise-runtimeerrorex.readthis-function-retrieves-a-single-entry-from-a-database-entryid-id-for-entry-e.g.mn908947-dbname-name-of-database-e.g.-genbankonce-we-got-that-dna-string-consisting-of-only-4-types-of-characters-actg-a-biologist-will-translate-them-into-aminoacid-or-protein-string-consisting-20-types-of-characters-asterisk-please-see-httpswww.hgvs.orgmutnomencodon.html.-well-translate-them-using-a-dictionary-code-not-shown-and-then-well-split-them-based-on-so-that-we-can-generate-an-output-of-a-list-of-strings-that-is-stored-in-a-variable-called-protseq-see-httpsdocs.google.comdocumentd1r22igmfe9i1tyalpk5zsoikv-on6xzvuq-c87xdmilgedituspsharing-finally-from-that-i-need-to-find-how-many-strings-inside-of-that-protseq-that-meet-these-criteria1-start-with-m2-end-with-3-has-length-of-100-charactersso-my-question-is-actuallyhow-to-understand-the-logic-behind-the-below-code-that-is-said-to-be-able-to-do-that-jobcheck-first-occurence-of-m-in-each-string-of-that-protseq-listwhere_m_in_protseq_1st-string-protseq0.findmprintwhere_m_in_protseq_1st-stringprintlenprotseq0code-to-check-whether-each-string-in-that-protseq-start-with-m-and-100cnt-0for-i-in-protseqif-leni---i.findm-100cnt-1printcnt-for-seq-in-protseqm_pos-seq.findm-m_end_seq-seqm_posif-lenm_end_seq-100printm_end_seqprintlenm_end_seq-the-answer-for-this-problem-should-be-8-strings-inside-18-members-of-that-protseq-list-will-meet-that-criteria-so-if-we-can-end-up-getting-8-from-that-code-we-are-correct-but-i-just-hope-that-i-can-get-it-correct-and-understand-how-the-code-worksimportance-from-that-8-strings-we-can-try-1-of-them-in-a-real-protein-database-called-uniprot-kb-and-know-what-part-of-covid-body-that-is-likely-to-interact-with-human-and-causing-diseasethank-you-very-much" class="nav-link" data-scroll-target="#question-1---finding-substrings-for-covid-sequencing---note-that-the-formatting-has-not-transferred-correctly-here-nidanida-has-a-problem-where-she-needs-to-identify-specific-substrings-from-a-large-sequence-of-characters-dna-sequence.-her-code-is-belowi-just-found-out-that-the-code-covid_seq-getsequencemn908947-genbank-dna_alphabet-is-actually-trying-to-get-a-dna-sequence-of-covid-from-genbank-will-give-output-of-a-string-containing-29900-characters-that-we-can-actually-see-by-clicking-on-the-ebi-website-below-it-is-use-a-function-called-getsequence-with-codedef-getsequenceentryid-dbname-uniprotkb-alphabet-protein_alphabet-format-fasta-debug-bool-true-if-not-isinstanceentryid-str-entryid-entryid.decodeutf-8-url-httpwww.ebi.ac.uktoolsdbfetchdbfetchstylerawdb-dbname-format-format-id-entryid-try-if-debug-printdebug-querying-url-0.formaturl-data-urllib.request.urlopenurl.read-if-format-fasta-return-readfastastringdata.decodeutf-8-alphabet0-else-return-data.decodeutf-8-except-urllib.error.httperror-as-ex-raise-runtimeerrorex.readthis-function-retrieves-a-single-entry-from-a-database-entryid-id-for-entry-e.g.mn908947-dbname-name-of-database-e.g.-genbankonce-we-got-that-dna-string-consisting-of-only-4-types-of-characters-actg-a-biologist-will-translate-them-into-aminoacid-or-protein-string-consisting-20-types-of-characters-asterisk-please-see-httpswww.hgvs.orgmutnomencodon.html.-well-translate-them-using-a-dictionary-code-not-shown-and-then-well-split-them-based-on-so-that-we-can-generate-an-output-of-a-list-of-strings-that-is-stored-in-a-variable-called-protseq-see-httpsdocs.google.comdocumentd1r22igmfe9i1tyalpk5zsoikv-on6xzvuq-c87xdmilgedituspsharing-finally-from-that-i-need-to-find-how-many-strings-inside-of-that-protseq-that-meet-these-criteria1-start-with-m2-end-with-3-has-length-of-100-charactersso-my-question-is-actuallyhow-to-understand-the-logic-behind-the-below-code-that-is-said-to-be-able-to-do-that-jobcheck-first-occurence-of-m-in-each-string-of-that-protseq-listwhere_m_in_protseq_1st-string-protseq0.findmprintwhere_m_in_protseq_1st-stringprintlenprotseq0code-to-check-whether-each-string-in-that-protseq-start-with-m-and-100cnt-0for-i-in-protseqif-leni---i.findm-100cnt-1printcnt-for-seq-in-protseqm_pos-seq.findm-m_end_seq-seqm_posif-lenm_end_seq-100printm_end_seqprintlenm_end_seq-the-answer-for-this-problem-should-be-8-strings-inside-18-members-of-that-protseq-list-will-meet-that-criteria-so-if-we-can-end-up-getting-8-from-that-code-we-are-correct-but-i-just-hope-that-i-can-get-it-correct-and-understand-how-the-code-worksimportance-from-that-8-strings-we-can-try-1-of-them-in-a-real-protein-database-called-uniprot-kb-and-know-what-part-of-covid-body-that-is-likely-to-interact-with-human-and-causing-diseasethank-you-very-much">Question 1 - Finding substrings for COVID sequencing - <strong>Note that the formatting has not transferred correctly here</strong> NidaNida has a problem where she needs to identify specific substrings from a large sequence of characters (DNA sequence). Her code is belowI just found out that the code: <code>covid_seq = getSequence('MN908947', 'genbank', DNA_Alphabet)</code> is actually trying to get a DNA sequence of covid from genbank &amp; will give output of a string containing 29,900 characters (that we can actually see by clicking on the EBI website below) it is use a function called <code>"getSequence"</code> with code:<code>def getSequence(entryId, dbName = 'uniprotkb', alphabet = Protein_Alphabet, format = 'fasta', debug: bool = True): if not isinstance(entryId, str): entryId = entryId.decode("utf-8") url ='http://www.ebi.ac.uk/Tools/dbfetch/dbfetch?style=raw&amp;db=' + dbName + '&amp;format=' + format + '&amp;id=' + entryId try: if debug: print('DEBUG: Querying URL: {0}'.format(url)) data = urllib.request.urlopen(url).read() if format == 'fasta': return readFastaString(data.decode("utf-8"), alphabet)[0] else: return data.decode("utf-8") except urllib.error.HTTPError as ex: raise RuntimeError(ex.read())</code>#This function retrieves a single entry from a database (entryId: ID for entry e.g.’MN908947’, dbName: name of database e.g.&nbsp;‘genbank’)once we got that DNA string (consisting of only 4 types of characters-A/C/T/G), a biologist will translate them (into AminoAcid or Protein string, consisting 20 types of characters + asterisk/* , please see https://www.hgvs.org/mutnomen/codon.html). We’ll translate them using a dictionary (code not shown) and then we’ll split them based on * so that we can generate an output of a list of strings that is stored in a variable called “protseq” (see https://docs.google.com/document/d/1R22IGMfe9i1tYAlPK5ZSOikV-ON6xZVUq-C87xDMiLg/edit?usp=sharing )finally, from that I need to find how many strings inside of that “protseq” that meet these criteria:1) start with M2) end with *3) has length of &gt;=100 charactersso, my question is actually:how to understand the logic behind the below code that is said to be able to do that job:#check first occurence of M in each string of that ‘protseq’ list&gt;&gt;where_M_in_protseq_1st-string = protseq[0].find(‘M’)print(where_M_in_protseq_1st-string)print(len(protseq[0]))#code to check whether each string in that ‘protseq’ start with M and &gt;= 100&gt;&gt;cnt = 0&gt;&gt;for i in protseq:&gt;&gt;&gt;if len(i) - i.find(“M”) &gt;= 100:&gt;&gt;&gt;&gt;cnt +=1&gt;&gt;print(cnt) &gt;&gt;for seq in protseq:&gt;&gt;&gt;m_pos = seq.find(“M”) &gt;&gt;&gt;m_end_seq = seq[m_pos:]&gt;&gt;&gt;if len(m_end_seq) &gt; 100:&gt;&gt;&gt;&gt;print(m_end_seq)&gt;&gt;&gt;&gt;print(len(m_end_seq)) The answer for this problem should be 8 strings inside 18 members of that protseq list will meet that criteria (so if we can end up getting 8 from that code we are correct)–&gt; (but I just hope that I can get it correct and understand how the code works)Importance: from that 8 strings, we can try 1 of them in a real protein database (called Uniprot-KB) and know what part of covid body that is likely to interact with human and causing diseaseThank you very much</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item">Session Notebooks</li><li class="breadcrumb-item"><a href="../../Session_notebooks/Past - 2023/1-Sep23.html">Past - 2023</a></li><li class="breadcrumb-item"><a href="../../Session_notebooks/Past - 2023/1-Sep23.html">2023-09-07: UQ PUG 1</a></li></ol></nav>
<div class="quarto-title">
<h1 class="title">2023-09-07: UQ PUG 1</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>Welcome to UQ Python User Group! Check out our <a href="https://app.noteable.io/f/c84cf34f-03eb-48eb-aff4-ccf80b013b5c/UQ-PUG-General-Info.ipynb?cellID=1e2cf356">general information</a> for details about who we are and what we do.## Structure1. We will start today by having everyone add their names to this page.1. Add your questions to this page.1. This Month’s Presentation.1. Finally we will spend the rest of the session answering the questions you have brought!</p>
<section id="this-months-presentationwelcome-to-our-first-python-user-group-gathering-this-month-luke-and-cameron-give-an-overview-of-the-group-our-vision-and-noteable-the-interactive-collaborative-notebook-platform-which-can-run-markdown-python-r-and-sql." class="level2">
<h2 class="anchored" data-anchor-id="this-months-presentationwelcome-to-our-first-python-user-group-gathering-this-month-luke-and-cameron-give-an-overview-of-the-group-our-vision-and-noteable-the-interactive-collaborative-notebook-platform-which-can-run-markdown-python-r-and-sql.">This month’s presentationWelcome to our first Python User Group gathering! This month, Luke and Cameron give an overview of the group, our vision and Noteable, the interactive collaborative notebook platform which can run markdown, Python, R and SQL.</h2>
</section>
<section id="introduce-yourself-whats-your-name-where-are-you-from-why-are-you-here-luke-library-learn-nick-library-learn-valentina-urrutia-guada-library-learn-nida-school-of-molecular-biology-1-understanding-the-logic-in-some-codes-for-bioinformatics-and-metagenomics-application-2-learn-how-to-visualize-data-in-python-that-is-hard-to-do-in-excel-cameron-library-learn-sam-hames-school-of-languages-and-cultures-communitynikhil-school-of-eecs-learn-paul-vrbik-eecs-support-jason-dail-senv-learn-annie-nguyen-senv-learn" class="level2">
<h2 class="anchored" data-anchor-id="introduce-yourself-whats-your-name-where-are-you-from-why-are-you-here-luke-library-learn-nick-library-learn-valentina-urrutia-guada-library-learn-nida-school-of-molecular-biology-1-understanding-the-logic-in-some-codes-for-bioinformatics-and-metagenomics-application-2-learn-how-to-visualize-data-in-python-that-is-hard-to-do-in-excel-cameron-library-learn-sam-hames-school-of-languages-and-cultures-communitynikhil-school-of-eecs-learn-paul-vrbik-eecs-support-jason-dail-senv-learn-annie-nguyen-senv-learn">Introduce yourself| What’s your name? | Where are you from? | Why are you here? | | — | — | — ||Luke |Library |Learn ||Nick |Library |Learn ||Valentina Urrutia Guada |Library |Learn ||Nida |School of Molecular Biology |(1) Understanding the logic (in some codes) for bioinformatics and metagenomics application, (2) Learn how to visualize data in python that is hard to do in excel || Cameron | Library | Learn ||Sam Hames | School of Languages and Cultures| Community||Nikhil |School of EECS | Learn|| Paul Vrbik | EECS | Support || Jason Dail| SENV | Learn ||Annie Nguyen |SENV | Learn :)|</h2>
</section>
<section id="research-toolshere-are-a-few-links-we-shared-around-mostly-from-jason.1.-to-assist-with-academic-citations1.-httpsresearchrabbitapp.comhome-helpful-when-making-research-collections-mapping-concepts-and-looking-at-the-linkages-between-references1.-httpsconsensus.app1.-httpschat.openai.com1.-httpsarticle-summarizer.scholarcy.comsummarizer1.-httpstypeset.io-ai-assistant-for-reading-and-understanding-papers1.-httpswww.listendata.com202303how-to-run-chatgpt-inside-excel.html-excel-extension-for-chat-gpt-how-to1.-eecs-tutor-help-httpseecs.uq.edu.aucurrent-studentseecs-learning-centre-tutors" class="level2">
<h2 class="anchored" data-anchor-id="research-toolshere-are-a-few-links-we-shared-around-mostly-from-jason.1.-to-assist-with-academic-citations1.-httpsresearchrabbitapp.comhome-helpful-when-making-research-collections-mapping-concepts-and-looking-at-the-linkages-between-references1.-httpsconsensus.app1.-httpschat.openai.com1.-httpsarticle-summarizer.scholarcy.comsummarizer1.-httpstypeset.io-ai-assistant-for-reading-and-understanding-papers1.-httpswww.listendata.com202303how-to-run-chatgpt-inside-excel.html-excel-extension-for-chat-gpt-how-to1.-eecs-tutor-help-httpseecs.uq.edu.aucurrent-studentseecs-learning-centre-tutors">Research toolsHere are a few links we shared around, mostly from Jason.1. <a href="https://scite.ai/">to assist with academic citations</a>1. https://researchrabbitapp.com/home helpful when making research collections, mapping concepts, and looking at the linkages between references1. https://consensus.app/1. https://chat.openai.com/1. https://article-summarizer.scholarcy.com/summarizer1. https://typeset.io/ AI assistant for reading and understanding papers1. https://www.listendata.com/2023/03/how-to-run-chatgpt-inside-excel.html (excel extension for chat GPT how-to)1. EECS tutor help: https://eecs.uq.edu.au/current-students/eecs-learning-centre-tutors</h2>
</section>
<section id="questionsif-you-have-any-python-questions-youd-like-to-explore-with-the-group-please-put-them-in-a-markdown-cell-with-any-code-youd-like-us-to-run-in-a-python-cell." class="level2">
<h2 class="anchored" data-anchor-id="questionsif-you-have-any-python-questions-youd-like-to-explore-with-the-group-please-put-them-in-a-markdown-cell-with-any-code-youd-like-us-to-run-in-a-python-cell.">QuestionsIf you have any Python questions you’d like to explore with the group, please put them in a markdown cell, with any code you’d like us to run in a Python cell.</h2>
<section id="question-1---finding-substrings-for-covid-sequencing---note-that-the-formatting-has-not-transferred-correctly-here-nidanida-has-a-problem-where-she-needs-to-identify-specific-substrings-from-a-large-sequence-of-characters-dna-sequence.-her-code-is-belowi-just-found-out-that-the-code-covid_seq-getsequencemn908947-genbank-dna_alphabet-is-actually-trying-to-get-a-dna-sequence-of-covid-from-genbank-will-give-output-of-a-string-containing-29900-characters-that-we-can-actually-see-by-clicking-on-the-ebi-website-below-it-is-use-a-function-called-getsequence-with-codedef-getsequenceentryid-dbname-uniprotkb-alphabet-protein_alphabet-format-fasta-debug-bool-true-if-not-isinstanceentryid-str-entryid-entryid.decodeutf-8-url-httpwww.ebi.ac.uktoolsdbfetchdbfetchstylerawdb-dbname-format-format-id-entryid-try-if-debug-printdebug-querying-url-0.formaturl-data-urllib.request.urlopenurl.read-if-format-fasta-return-readfastastringdata.decodeutf-8-alphabet0-else-return-data.decodeutf-8-except-urllib.error.httperror-as-ex-raise-runtimeerrorex.readthis-function-retrieves-a-single-entry-from-a-database-entryid-id-for-entry-e.g.mn908947-dbname-name-of-database-e.g.-genbankonce-we-got-that-dna-string-consisting-of-only-4-types-of-characters-actg-a-biologist-will-translate-them-into-aminoacid-or-protein-string-consisting-20-types-of-characters-asterisk-please-see-httpswww.hgvs.orgmutnomencodon.html.-well-translate-them-using-a-dictionary-code-not-shown-and-then-well-split-them-based-on-so-that-we-can-generate-an-output-of-a-list-of-strings-that-is-stored-in-a-variable-called-protseq-see-httpsdocs.google.comdocumentd1r22igmfe9i1tyalpk5zsoikv-on6xzvuq-c87xdmilgedituspsharing-finally-from-that-i-need-to-find-how-many-strings-inside-of-that-protseq-that-meet-these-criteria1-start-with-m2-end-with-3-has-length-of-100-charactersso-my-question-is-actuallyhow-to-understand-the-logic-behind-the-below-code-that-is-said-to-be-able-to-do-that-jobcheck-first-occurence-of-m-in-each-string-of-that-protseq-listwhere_m_in_protseq_1st-string-protseq0.findmprintwhere_m_in_protseq_1st-stringprintlenprotseq0code-to-check-whether-each-string-in-that-protseq-start-with-m-and-100cnt-0for-i-in-protseqif-leni---i.findm-100cnt-1printcnt-for-seq-in-protseqm_pos-seq.findm-m_end_seq-seqm_posif-lenm_end_seq-100printm_end_seqprintlenm_end_seq-the-answer-for-this-problem-should-be-8-strings-inside-18-members-of-that-protseq-list-will-meet-that-criteria-so-if-we-can-end-up-getting-8-from-that-code-we-are-correct-but-i-just-hope-that-i-can-get-it-correct-and-understand-how-the-code-worksimportance-from-that-8-strings-we-can-try-1-of-them-in-a-real-protein-database-called-uniprot-kb-and-know-what-part-of-covid-body-that-is-likely-to-interact-with-human-and-causing-diseasethank-you-very-much" class="level3">
<h3 class="anchored" data-anchor-id="question-1---finding-substrings-for-covid-sequencing---note-that-the-formatting-has-not-transferred-correctly-here-nidanida-has-a-problem-where-she-needs-to-identify-specific-substrings-from-a-large-sequence-of-characters-dna-sequence.-her-code-is-belowi-just-found-out-that-the-code-covid_seq-getsequencemn908947-genbank-dna_alphabet-is-actually-trying-to-get-a-dna-sequence-of-covid-from-genbank-will-give-output-of-a-string-containing-29900-characters-that-we-can-actually-see-by-clicking-on-the-ebi-website-below-it-is-use-a-function-called-getsequence-with-codedef-getsequenceentryid-dbname-uniprotkb-alphabet-protein_alphabet-format-fasta-debug-bool-true-if-not-isinstanceentryid-str-entryid-entryid.decodeutf-8-url-httpwww.ebi.ac.uktoolsdbfetchdbfetchstylerawdb-dbname-format-format-id-entryid-try-if-debug-printdebug-querying-url-0.formaturl-data-urllib.request.urlopenurl.read-if-format-fasta-return-readfastastringdata.decodeutf-8-alphabet0-else-return-data.decodeutf-8-except-urllib.error.httperror-as-ex-raise-runtimeerrorex.readthis-function-retrieves-a-single-entry-from-a-database-entryid-id-for-entry-e.g.mn908947-dbname-name-of-database-e.g.-genbankonce-we-got-that-dna-string-consisting-of-only-4-types-of-characters-actg-a-biologist-will-translate-them-into-aminoacid-or-protein-string-consisting-20-types-of-characters-asterisk-please-see-httpswww.hgvs.orgmutnomencodon.html.-well-translate-them-using-a-dictionary-code-not-shown-and-then-well-split-them-based-on-so-that-we-can-generate-an-output-of-a-list-of-strings-that-is-stored-in-a-variable-called-protseq-see-httpsdocs.google.comdocumentd1r22igmfe9i1tyalpk5zsoikv-on6xzvuq-c87xdmilgedituspsharing-finally-from-that-i-need-to-find-how-many-strings-inside-of-that-protseq-that-meet-these-criteria1-start-with-m2-end-with-3-has-length-of-100-charactersso-my-question-is-actuallyhow-to-understand-the-logic-behind-the-below-code-that-is-said-to-be-able-to-do-that-jobcheck-first-occurence-of-m-in-each-string-of-that-protseq-listwhere_m_in_protseq_1st-string-protseq0.findmprintwhere_m_in_protseq_1st-stringprintlenprotseq0code-to-check-whether-each-string-in-that-protseq-start-with-m-and-100cnt-0for-i-in-protseqif-leni---i.findm-100cnt-1printcnt-for-seq-in-protseqm_pos-seq.findm-m_end_seq-seqm_posif-lenm_end_seq-100printm_end_seqprintlenm_end_seq-the-answer-for-this-problem-should-be-8-strings-inside-18-members-of-that-protseq-list-will-meet-that-criteria-so-if-we-can-end-up-getting-8-from-that-code-we-are-correct-but-i-just-hope-that-i-can-get-it-correct-and-understand-how-the-code-worksimportance-from-that-8-strings-we-can-try-1-of-them-in-a-real-protein-database-called-uniprot-kb-and-know-what-part-of-covid-body-that-is-likely-to-interact-with-human-and-causing-diseasethank-you-very-much">Question 1 - Finding substrings for COVID sequencing - <strong>Note that the formatting has not transferred correctly here</strong> NidaNida has a problem where she needs to identify specific substrings from a large sequence of characters (DNA sequence). Her code is belowI just found out that the code: <code>covid_seq = getSequence('MN908947', 'genbank', DNA_Alphabet)</code> is actually trying to get a DNA sequence of covid from genbank &amp; will give output of a string containing 29,900 characters (that we can actually see by clicking on the EBI website below) it is use a function called <code>"getSequence"</code> with code:<code>def getSequence(entryId, dbName = 'uniprotkb', alphabet = Protein_Alphabet, format = 'fasta', debug: bool = True): if not isinstance(entryId, str): entryId = entryId.decode("utf-8") url ='http://www.ebi.ac.uk/Tools/dbfetch/dbfetch?style=raw&amp;db=' + dbName + '&amp;format=' + format + '&amp;id=' + entryId try: if debug: print('DEBUG: Querying URL: {0}'.format(url)) data = urllib.request.urlopen(url).read() if format == 'fasta': return readFastaString(data.decode("utf-8"), alphabet)[0] else: return data.decode("utf-8") except urllib.error.HTTPError as ex: raise RuntimeError(ex.read())</code>#This function retrieves a single entry from a database (entryId: ID for entry e.g.’MN908947’, dbName: name of database e.g.&nbsp;‘genbank’)once we got that DNA string (consisting of only 4 types of characters-A/C/T/G), a biologist will translate them (into AminoAcid or Protein string, consisting 20 types of characters + asterisk/* , please see https://www.hgvs.org/mutnomen/codon.html). We’ll translate them using a dictionary (code not shown) and then we’ll split them based on * so that we can generate an output of a list of strings that is stored in a variable called “protseq” (see https://docs.google.com/document/d/1R22IGMfe9i1tYAlPK5ZSOikV-ON6xZVUq-C87xDMiLg/edit?usp=sharing )finally, from that I need to find how many strings inside of that “protseq” that meet these criteria:1) start with M2) end with *3) has length of &gt;=100 charactersso, my question is actually:how to understand the logic behind the below code that is said to be able to do that job:#check first occurence of M in each string of that ‘protseq’ list&gt;&gt;where_M_in_protseq_1st-string = protseq[0].find(‘M’)print(where_M_in_protseq_1st-string)print(len(protseq[0]))#code to check whether each string in that ‘protseq’ start with M and &gt;= 100&gt;&gt;cnt = 0&gt;&gt;for i in protseq:&gt;&gt;&gt;if len(i) - i.find(“M”) &gt;= 100:&gt;&gt;&gt;&gt;cnt +=1&gt;&gt;print(cnt) &gt;&gt;for seq in protseq:&gt;&gt;&gt;m_pos = seq.find(“M”) &gt;&gt;&gt;m_end_seq = seq[m_pos:]&gt;&gt;&gt;if len(m_end_seq) &gt; 100:&gt;&gt;&gt;&gt;print(m_end_seq)&gt;&gt;&gt;&gt;print(len(m_end_seq)) The answer for this problem should be 8 strings inside 18 members of that protseq list will meet that criteria (so if we can end up getting 8 from that code we are correct)–&gt; (but I just hope that I can get it correct and understand how the code works)Importance: from that 8 strings, we can try 1 of them in a real protein database (called Uniprot-KB) and know what part of covid body that is likely to interact with human and causing diseaseThank you very much</h3>
<div id="06eb4a2b-b9f4-48d2-954d-837010fcebc4" class="cell" data-noteable="{&quot;cell_type&quot;:&quot;python&quot;,&quot;output_collection_id&quot;:&quot;2c1d94a6-de50-41eb-b9b3-111644966291&quot;}" data-tags="[&quot;show_line_numbers&quot;]" data-executetime="{&quot;end_time&quot;:&quot;2023-09-07T01:20:37.745011+00:00&quot;,&quot;start_time&quot;:&quot;2023-09-07T01:20:37.021769+00:00&quot;}" data-jupyter="{&quot;source_hidden&quot;:true}">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> urllibclass Alphabet():    <span class="st">""" A minimal class for alphabets        Alphabets include DNA, RNA and Protein """</span>    <span class="kw">def</span> <span class="fu">__init__</span>(<span class="va">self</span>, symbolString):        <span class="va">self</span>.symbols <span class="op">=</span> symbolString    <span class="kw">def</span> <span class="fu">__len__</span>(<span class="va">self</span>):              <span class="co"># implements the "len" operator, e.g. "len(Alphabet('XYZ'))" results in 3        return len(self.symbols)    # will tell you the length of the symbols in an Alphabet instance    def __contains__(self, sym):    # implements the "in" operator, e.g. "'A' in Alphabet('ACGT')" results in True        return sym in self.symbols  # will tell you if 'A' is in the symbols in an Alphabet instance    def __iter__(self):             # method that allows us to iterate over all symbols, e.g. "for sym in Alphabet('ACGT'): print sym" prints A, C, G and T on separate lines        tsyms = tuple(self.symbols)        return tsyms.__iter__()    def __getitem__(self, ndx):        """ Retrieve the symbol(s) at the specified index (or slice of indices) """        return self.symbols[ndx]    def index(self, sym):        """ Retrieve the index of the given symbol in the alphabet. """        return self.symbols.index(sym)    def __str__(self):        return self.symbols""" Below we declare alphabet variables that are going to be available whenthis module (this .py file) is imported """DNA_Alphabet = Alphabet('ACGT')RNA_Alphabet = Alphabet('ACGU')Protein_Alphabet = Alphabet('ACDEFGHIKLMNPQRSTVWY')Protein_wX = Alphabet('ACDEFGHIKLMNPQRSTVWYX')Protein_wGAP = Alphabet('ACDEFGHIKLMNPQRSTVWY-')def getSequence(entryId, dbName = 'uniprotkb', alphabet = Protein_Alphabet, format = 'fasta', debug: bool = True):    """ Retrieve a single entry from a database    entryId: ID for entry e.g. 'P63166' or 'SUMO1_MOUSE'    dbName: name of database e.g. 'uniprotkb' or 'pdb' or 'refseqn'; see http://www.ebi.ac.uk/Tools/dbfetch/dbfetch/dbfetch.databases for available databases    format: file format specific to database e.g. 'fasta' or 'uniprot' for uniprotkb (see http://www.ebi.ac.uk/Tools/dbfetch/dbfetch/dbfetch.databases)    See http://www.ebi.ac.uk/Tools/dbfetch/syntax.jsp for more info re URL syntax    """    if not isinstance(entryId, str):        entryId = entryId.decode("utf-8")    url ='http://www.ebi.ac.uk/Tools/dbfetch/dbfetch?style=raw&amp;db=' + dbName + '&amp;format=' + format + '&amp;id=' + entryId    try:        if debug:            print('DEBUG: Querying URL: {0}'.format(url))        data = urllib.request.urlopen(url).read()        if format == 'fasta':            return readFastaString(data.decode("utf-8"), alphabet)[0]        else:            return data.decode("utf-8")    except urllib.error.HTTPError as ex:        raise RuntimeError(ex.read())# get the covid 19 genome (29kB)covid_seq = getSequence('MN908947', 'genbank', DNA_Alphabet)# print(seq_no10)# get all the bases from covid_seq# translate all of those base into amino acid seq# in all reading frames (6)tr_f = [0, 1, 2]# translate proteincovid_AAseq = []for i in tr_f:    #print("all ORF in fwd direction", covid_seq.translateDNA(i, True))    seq10b=covid_seq.translateDNA(i, True)    protseq=seq10b.split("*")    seq10br = covid_seq.translateDNA(i, False)    protseq.extend(seq10br.split("*"))print(str(protseq))#     print(protseq)#     for element in protseq:#         print("Individual value is",element)# #for i in tr_f:#     print("all ORF in reverse direction", covid_seq.translateDNA(i, False))    #NEXT STEP (for mapping the ORF that begins with M &amp; calculate the len &gt; 100)ORF = []for i in protseq:    if i.startswith('M') == True:        ORF.append(i)print('all of potential ORF', ORF)print('length of all potential ORF')for i in ORF:    print(i, ":", len(i))true_ORF = []for i in ORF:    if len(i) &gt; 100:        true_ORF.append(i)print(true_ORF)# for i in protseq:#     if i == 'M': # and len(i) &gt; 100:#         ORF.append(i)# print(ORF)        </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="41454b45-b706-4f41-b1c0-dabc82e17d58" class="cell" data-noteable="{&quot;cell_type&quot;:&quot;python&quot;,&quot;output_collection_id&quot;:&quot;2f775554-18ba-44fa-935d-af5a5ee57969&quot;}" data-executetime="{&quot;end_time&quot;:&quot;2023-09-07T01:35:28.222291+00:00&quot;,&quot;start_time&quot;:&quot;2023-09-07T01:35:28.067202+00:00&quot;}">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>characters <span class="op">=</span> <span class="st">"asjidowgeriogpjicnjjlaksdjalksdj*alskjdjjjjeosj   hjjjjl"</span><span class="co"># We want to pick out "jid" and "jic"ans = []for k, c in enumerate(characters):    if c == "j":   # add abtrary number of constraints        ans.append(characters[k:k+3])    if c == "*":        breakif len(ans[-1]) &lt; 3:    ans.pop()        print(ans)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cf536d8f-7335-4017-8306-23af57d35577" class="cell" data-jupyter="{&quot;outputs_hidden&quot;:false,&quot;source_hidden&quot;:false}" data-noteable="{&quot;cell_type&quot;:&quot;code&quot;}">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>characters <span class="op">=</span> <span class="st">"asjidowgeriogpjicneosjl"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/uqpug\.github\.io\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>